libs/@local/graph/migrations-macros/src/lib.rs

```diff
@@ -0,0 +1,370 @@
+extern crate alloc;
+
+use alloc::collections::{BTreeMap, btree_map::Entry};
+use core::{error::Error, num::ParseIntError};
+use std::{
+    fs, io,
+    path::{Path, PathBuf},
+    sync::OnceLock,
+};
+
+use convert_case::{Case, Casing as _};
+use proc_macro2::TokenStream;
+use quote::{format_ident, quote};
+use regex::Regex;
+use sha2::{Digest as _, Sha256};
+use syn::{LitStr, parse_macro_input};
+use walkdir::{DirEntry, WalkDir};
+
+pub(crate) fn crate_root() -> PathBuf {
+    PathBuf::from(std::env::var("CARGO_MANIFEST_DIR").expect("Could not read crate root"))
+}
+
+#[derive(Debug, Clone)]
+struct MigrationFile {
+    module_path: PathBuf,
+    name: String,
+    number: u32,
+    size: usize,
+    digest: [u8; 32],
+}
+
+fn read_directory_content(path: impl AsRef<Path>) -> io::Result<String> {
+    WalkDir::new(path)
+        .sort_by_file_name()
+        .into_iter()
+        // .filter(|entry| Ok::<_, io::Error>(!entry?.file_type().is_dir()))
+        .try_fold(String::new(), |mut buffer, entry| {
+            let entry = entry?;
+            if !entry.file_type().is_dir()  {
+                buffer.push_str(&fs::read_to_string(entry.path())?);
+            }
+            Ok(buffer)
+        })
+}
+
+const BASE_REGEX: &str = r"^v(\d+)__(\w+)";
+static DIRECTORY_REGEX: OnceLock<Regex> = OnceLock::new();
```

Comment:
isn't a [`LazyLock`](https://doc.rust-lang.org/stable/std/sync/struct.LazyLock.html) more appropriate here?

Reply:
That's true, thanks!

Reply:
Did as drive-by in [`17d3483` (#5674)](https://github.com/hashintel/hash/pull/5674/commits/17d3483303950307d995d5daf13a09793f013b89)